v4/export/dump.go

```diff
@@ -114,164 +117,384 @@ func (d *Dumper) Dump() (err error) {
+	conf := d.conf
 	allTables := conf.Tables
-	g, ctx := errgroup.WithContext(pCtx)
-	tableDataIRTotal := make([]TableDataIR, 0, len(allTables))
-	splitChunkStart := time.Now()
 	for dbName, tables := range allTables {
-		createDatabaseSQL, err := ShowCreateDatabase(connectPool.extraConn(), dbName)
+		createDatabaseSQL, err := ShowCreateDatabase(metaConn, dbName)
 		if err != nil {
 			return err
 		}
-		if err := writer.WriteDatabaseMeta(ctx, dbName, createDatabaseSQL); err != nil {
-			return err
-		}
+		task := NewTaskDatabaseMeta(dbName, createDatabaseSQL)
+		d.sendTaskToChan(task, taskChan)
 
-		if len(tables) == 0 {
-			continue
-		}
 		for _, table := range tables {
-			table := table
-			tableDataIRArray, err := dumpTable(ctx, conf, connectPool.extraConn(), dbName, table, writer)
+			log.Debug("start dumping table...",
+				zap.String("table", table.Name))
+			meta, err := dumpTableMeta(conf, metaConn, dbName, table)
+			if err != nil {
+				return err
+			}
+
+			if table.Type == TableTypeView {
+				task := NewTaskViewMeta(dbName, table.Name, meta.ShowCreateTable(), meta.ShowCreateView())
+				d.sendTaskToChan(task, taskChan)
+			} else {
+				task := NewTaskTableMeta(dbName, table.Name, meta.ShowCreateTable())
+				d.sendTaskToChan(task, taskChan)
+			}
+			err = d.dumpTableData(metaConn, meta, taskChan)
 			if err != nil {
 				return err
 			}
-			tableDataIRTotal = append(tableDataIRTotal, tableDataIRArray...)
 		}
 	}
-	summary.CollectDuration("split chunks", time.Since(splitChunkStart))
-	progressPrinter := utils.StartProgress(ctx, "dumpling", int64(len(tableDataIRTotal)), shouldRedirectLog(conf), log.Info)
-	defer progressPrinter.Close()
-	tableDataStartTime := time.Now()
-	for _, tableIR := range tableDataIRTotal {
-		tableIR := tableIR
-		g.Go(func() error {
-			conn := connectPool.getConn()
-			defer func() {
-				connectPool.releaseConn(conn)
-			}()
-			retryTime := 0
-			var lastErr error
-			return utils.WithRetry(ctx, func() (err error) {
-				defer func() {
-					lastErr = err
-					if err == nil {
-						progressPrinter.Inc()
-					} else {
-						errorCount.With(conf.Labels).Inc()
-					}
-				}()
-				retryTime += 1
-				log.Debug("trying to dump table chunk", zap.Int("retryTime", retryTime), zap.String("db", tableIR.DatabaseName()),
-					zap.String("table", tableIR.TableName()), zap.Int("chunkIndex", tableIR.ChunkIndex()), zap.NamedError("lastError", lastErr))
-				if retryTime > 1 {
-					conn, err = rebuildConnFunc(conn)
-					if err != nil {
-						return
-					}
-				}
-				err = tableIR.Start(ctx, conn)
-				if err != nil {
-					return
-				}
-				return writer.WriteTableData(ctx, tableIR)
-			}, newDumpChunkBackoffer(canRebuildConn(conf.Consistency, conf.TransactionalConsistency)))
-		})
+	return nil
+}
+
+func (d *Dumper) dumpTableData(conn *sql.Conn, meta TableMeta, taskChan chan<- Task) error {
+	conf := d.conf
+	if conf.NoData {
+		return nil
+	}
+	if conf.Rows == UnspecifiedSize {
```

Comment:
Will we support specifying rows for `tablesample` in the future? If not, specifying useless `rows` to enable dumping through `tablesameple` looks weird.

Reply:
Yes. Later we should also add some tests for this feature.